doc/dirvish-locate.1
.\" Automatically generated by Pod::Man 2.1801 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DIRVISH-LOCATE 1"
.TH DIRVISH-LOCATE 1 "2009-01-17" "perl v5.10.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dirvish\-locate \- locate file versions in dirvish images
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdirvish-locate\fR \fIvault\fR[\fI:branch\fR] \fIpattern\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Locate versions of files in a dirvish vault
.PP
The index of each image specified \fIvault\fR is searched for paths
matching \fIpattern\fR. Each path found matching the \fIpattern\fR will be
reported followed by a modification time of each version of the file
and all images having a link to it.
.PP
The optional \fIbranch\fR specification will restrict searching to the
specified branch.
.PP
Images with an error status will be skipped as will any without index
files. The index file may be compressed by gzip or bzip2. See \fBtree\fR
and \fBindex\fR in \fB\f(BIdirvish.conf\fB\|(5)\fR for details.
.PP
The \fIpattern\fR is a \fBperl\fR regular expression to match the final
component of the path. Append \fB.*\fR to the end of the pattern if you
wish to match any substring of the whole path or \fB$\fR if you wish to
anchor the pattern to the end of the path. See \fB\f(BIperlre\fB\|(1)\fR for
details.
.PP
Directories are excluded from matching as they would wind up matching
every file within them anyway. Symlinks are also excluded from
matching.
.PP
If the \fIpattern\fR matches too many paths \fBdirvish-locate\fR will only
report the paths matched and not versions. As a sanity check if the
number of matches is really excessive \fBdirvish-locate\fR will limit the
number of images searched. Excessive matches is an indication of an
insufficiently specific \fIpattern\fR. Use the resulting path list to
compose a more specific one.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
To facilitate further automation and integration of \fBdirvish-locate\fR
with other tools \fBdirvish-locate provides rationalized exit codes.\fR
The exit codes are range based. While the code for a specific error may
change from one version to another it will remain within the specified
range. So don't test for specific exit codes but instead test for a
range of values. To the degree possible higher value ranges indicate
more severe errors.
.IP "\fB0\fR" 10
.IX Item "0"
success 200\-219 An error was encountered in loading a configuration
file.
.IP "\fB220\-254\fR" 10
.IX Item "220-254"
An error was detected in the configuration.
.IP "\fB255\fR" 10
.IX Item "255"
Incorrect usage.
.SH "FILES"
.IX Header "FILES"
\&\fB/etc/dirvish/master.conf\fR
.PP
alternate master configuration file.
.PP
\&\fB/etc/dirvish.conf\fR
.PP
master configuration file.
.PP
\&\fIbank/vault/image/\fR\fBsummary\fR
.PP
image creation summary.
.PP
\&\fIbank/vault/image/\fR\fBindex\fR
.PP
\&\fIbank/vault/image/\fR\fBindex.gz\fR
.PP
\&\fIbank/vault/image/\fR\fBindex.bz2\fR
.PP
dirvish index file.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 1
\& dirvish.conf(5)
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Dirvish was created by J.W. Schultz of Pegasystems Technologies.
.PP
Dirvish is now maintained by Keith Lofstrom at www.dirvish.org , with
the able assistance of many others.
.SH "BUGS"
.IX Header "BUGS"


debian/changelog
dirvish (1.3.2-r657-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Sun, 08 Feb 2009 00:41:43 +0100

dirvish (1.3.2-r652-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 04 Feb 2009 19:34:46 +0100

dirvish (1.3.2-r634-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 28 Jan 2009 17:46:52 +0100

dirvish (1.3.2-r632-1) unstable; urgency=low

  * Initial release

 -- Dominik Schulz <dominik.schulz@gauner.org>  Tue, 27 Jan 2009 16:18:14 +0100


dirvish-locate.pl
#!/usr/bin/perl
# dirvish-locate
# 1.3.X series
# Copyright 2005 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 660 $
# Revision date   : $Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-locate.pl $
#
#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################
#
#----------------------------------------------------------------------------
# Revision information
#----------------------------------------------------------------------------
my %CodeID = (
    Rev    => '$Rev: 660 $'     ,
    Date   => '$Date: 2009-02-17 18:43:32 +0100 (Di, 17 Feb 2009) $'    ,
    Author => '$Author: tex $'  ,
    URL    => '$HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-locate.pl $' ,
);

$VERSION =   $CodeID{URL};
$VERSION =~  s#^.*dirvish_##;  # strip off the front
$VERSION =~  s#\/.*##;         # strip off the rear after the last /
$VERSION =~  s#[_-]#.#g;       # _ or - to "."

#----------------------------------------------------------------------------
# Modules and includes
#----------------------------------------------------------------------------
use strict;
use warnings;

use Time::ParseDate;
use POSIX qw(strftime);
use Getopt::Long;
use Dirvish;

use File::Spec::Functions;

#----------------------------------------------------------------------------
# SIG Handler
#----------------------------------------------------------------------------
$SIG{TERM} = \&sigterm; # handles "kill <PID>"
$SIG{INT} = \&sigterm; # handles Ctrl+C or "kill -2 <PID>"

#----------------------------------------------------------------------------
# Initialisation
#----------------------------------------------------------------------------
my $KILLCOUNT = 1000;
my $MAXCOUNT = 100;

my $Options = reset_options( \&usage, @ARGV);	# initialize the %$Options hash
load_master_config('f', $Options);				# load master config into $Options

GetOptions($Options, qw(
    version
    help|?
    )) or &usage();

my $Vault = shift;
my $Branch = undef;
$Vault =~ /:/ and ($Vault, $Branch) = split(/:/, $Vault);
my $Pattern = shift;

$Vault && length($Pattern) or &usage();
# prepend dot if asterisk or question mark is
# the first character. Make rsync-like patterns like *.xml work
$Pattern = ".".$Pattern if($Pattern =~ m/^(\*|\?)/);

my $fullpattern = $Pattern;
my $partpattern = undef;
$fullpattern =~ /\$$/ or $fullpattern .= '[^/]*$';
($partpattern = $fullpattern) =~ s/^\^//;

my $bank = undef;
for $b (@{$$Options{bank}})
{
    -d catdir($b,$Vault) and $bank = $b;
}
$bank or seppuku 220, "No such vault: $Vault";

opendir VAULT, catdir($bank,$Vault) or seppuku 221, "cannot open vault: $Vault";
my @invault = readdir(VAULT);
closedir VAULT;

my @images = ();
for my $image (@invault)
{
    $image eq 'dirvish' and next;
    my $imdir = catdir($bank,$Vault,$image);
    -f catfile($imdir,"summary") or next;
    (-l $imdir && $imdir =~ /current/) and next; # skip current-symlink
    my $conf = loadconfig('R', catfile($imdir,"summary"), $Options) or next;
    $$conf{Status} eq 'success' || $$conf{Status} =~ /^warn/
        or next;
    $$conf{'Backup-complete'} or next;
    $Branch && $$conf{branch} ne $Branch and next;
	
    unshift @images, {
        imdir   => $imdir,
        image   => $$conf{Image},
        branch  => $$conf{branch},
        created => $$conf{'Backup-complete'},
    }
}

my $imagecount = 0;
my $pathcount = 0;
my $path = undef;
my %match = ();
for my $image (sort(imsort_locate @images))
{
    my $imdir = $$image{imdir};

    my $index = undef;
    -f catfile($imdir,"index.bz2") and $index = "bzip2 -d -c ".catfile($imdir,"index.bz2")."|";
    -f catfile($imdir,"index.gz") and $index = "gzip -d -c ".catfile($imdir,"index.gz")."|";
    -f catfile($imdir,"index") and $index = "<".catfile($imdir,"index");
    $index or next;

    ++$imagecount;

	# can't use three-fold open here, see above
    open(INDEX, $index) or next;
    while (<INDEX>)
    {
        chomp;

        m($partpattern) or next;
		# this parse operation is too slow.  It might be faster as a
		# split with trimmed leading whitespace and remerged modtime
        my $f = { image => $image };
        (
             $$f{inode},
            $$f{blocks},
            $$f{perms},
            $$f{links},
            $$f{owner},
            $$f{group},
            $$f{bytes},
            $$f{mtime},
            $path
        ) = m<^
            \s*(\S+)        # inode
            \s+(\S+)        # block count
            \s+(\S+)        # perms
            \s+(\S+)        # link count
            \s+(\S+)        # owner
            \s+(\S+)        # group
            \s+(\S+)        # byte count
            \s+(\S+\s+\S+\s+\S+)    # date
            \s+(\S.*)        # path
        $>x;
        $$f{perms} =~ /^[dl]/ and next;
        $path =~ m($fullpattern) or next;

        exists($match{$path}) or ++$pathcount;
        push @{$match{$path}}, $f;
    }
    if ($pathcount >= $KILLCOUNT)
    {
        print "dirvish-locate: too many paths match pattern, interrupting search\n";
        last;
    }
}

printf "%d matches in %d images\n", $pathcount, $imagecount;

$pathcount >= $MAXCOUNT
    and printf "Pattern '%s' too vague, listing paths only.\n", $Pattern;

my $last = undef;
my $linesize = 0;
for my $path (sort(keys(%match)))
{
    $last = undef;
    print $path;

    if ($pathcount >= $MAXCOUNT)
    {
        print "\n";
               next;
    }

    for my $hit (@{$match{$path}})
    {
        my $inode = $$hit{inode};
        my $mtime = $$hit{mtime};
        my $image = $$hit{image}{image};
        if (defined($last) && $inode ne $last)
        {
            $linesize = 5 + length($mtime) + length($image);
            printf "\n    %s %s", $mtime, $image;
        } else {
            $linesize += length($image) + 2;
            if ($linesize > 78)
            {
                $linesize = 5 + length($mtime) + length($image);
                print "\n",
                    " " x (5 + length($mtime)),
                    $image;
            } else {
                printf ", %s", $$hit{image}{image};
            }
        }
        $last = $inode;
    }
    print "\n\n";
}

exit 0;
#----------------------------------------------------------------------------
# Subs
#----------------------------------------------------------------------------
# Sort images
sub imsort_locate {
	## WARNING:  don't mess with the sort order, it is needed so that if
	## WARNING:  all images are expired the newest will be retained.
	$$a{branch} cmp $$b{branch}
	  || $$a{created} cmp $$b{created};
}
sub usage
{
    my $message = shift(@_);

    length($message) and print STDERR $message, "\n\n";

    print STDERR <<EOUSAGE;
USAGE
	dirvish-locate vault[:branch] pattern
	
	Pattern can be any PCRE.
	
EOUSAGE
	exit 255;
}
# Handle SIGTERM (SIG-15)
sub sigterm
{
	print STDERR "Received SIGTERM. Aborting running backup ...";
	# kill childs - kill(TERM, -$$):
	use POSIX;
	my $cnt = kill(SIGTERM, -$$);
	no POSIX;
	print STDERR "Signaled $cnt processes in current processgroup";
	# quit
	exit;
}
#----------------------------------------------------------------------------
# EOF
#----------------------------------------------------------------------------